vector.js

```diff
@@ -1,212 +1,283 @@
-        if (_.isUndefined(theta)) {
-            theta = radius[1];
-            radius = radius[0];
-        }
-
-        return kvector.cartFromPolarRad(radius, theta * Math.PI / 180);
-    },
-
-    // Rotate vector
-    rotateRad: function(v, theta) {
-        var polar = kvector.polarRadFromCart(v);
-        var angle = polar[1] + theta;
-        return kvector.cartFromPolarRad(polar[0], angle);
-    },
-
-    rotateDeg: function(v, theta) {
-        var polar = kvector.polarDegFromCart(v);
-        var angle = polar[1] + theta;
-        return kvector.cartFromPolarDeg(polar[0], angle);
-    },
-
-    // Angle between two vectors
-    angleRad: function(v1, v2) {
-        return Math.acos(kvector.dot(v1, v2) /
-            (kvector.length(v1) * kvector.length(v2)));
-    },
-
-    angleDeg: function(v1, v2) {
-        return kvector.angleRad(v1, v2) * 180 / Math.PI;
-    },
-
-    // Vector projection of v1 onto v2
-    projection: function(v1, v2) {
-        var scalar = kvector.dot(v1, v2) / kvector.dot(v2, v2);
-        return kvector.scale(v2, scalar);
-    },
-
-    // Round each number to a certain number of decimal places
-    round: function(vec, precision) {
-        return _.map(vec, function(elem, i) {
-            return knumber.round(elem, precision[i] || precision);
-        });
-    },
-
-    // Round each number to the nearest increment
-    roundTo: function(vec, increment) {
-        return _.map(vec, function(elem, i) {
-            return knumber.roundTo(elem, increment[i] || increment);
-        });
-    },
-
-    floorTo: function(vec, increment) {
-        return _.map(vec, function(elem, i) {
-            return knumber.floorTo(elem, increment[i] || increment);
-        });
-    },
-
-    ceilTo: function(vec, increment) {
-        return _.map(vec, function(elem, i) {
-            return knumber.ceilTo(elem, increment[i] || increment);
-        });
+import _ from "underscore";
+import * as knumber from "./number.js";
+
+function arraySum(array: $ReadOnlyArray<number>): number {
+    return array.reduce((memo, arg) => memo + arg, 0);
+}
+
+function arrayProduct(array: $ReadOnlyArray<number>): number {
+    return array.reduce((memo, arg) => memo * arg, 1);
+}
+
+export function is<T>(vec: $ReadOnlyArray<T>, dimension: number): boolean {
+    if (!_.isArray(vec)) {
+        return false;
+    }
+    if (dimension !== undefined && vec.length !== dimension) {
+        return false;
+    }
+    return vec.every(knumber.is);
+}
+
+// Normalize to a unit vector
+export function normalize(v: $ReadOnlyArray<number>): $ReadOnlyArray<number> {
+    return scale(v, 1 / length(v));
+}
+
+// Length/magnitude of a vector
+export function length(v: $ReadOnlyArray<number>): number {
+    return Math.sqrt(dot(v, v));
+}
+// Dot product of two vectors
+export function dot(
+    a: $ReadOnlyArray<number>,
+    b: $ReadOnlyArray<number>
+): number {
+    var vecs = _.toArray(arguments);
+    var zipped = _.zip.apply(_, vecs);
```

Comment:
I'm not sure why we using `_.toArray(arguments)` here given that the number of args is known.  This could be replaced with:
```
const zipped = _.zip(a, b);
```